### DISCUSSION TITLE
base64 hiccup

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12013] [Using MicroPython]
When I base64 encode b'0067462314' I get 'MDA2NzQ2MjMxNA==' but a web encoder (also claiming to be base 64) gives 'AgdGIxQ='. I can't work out how to get this in upython or python. Can someone please point me in the right direction?

**Comment by GitHubsSilverBullet** [comment]
```python
~$ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import base64
>>> base64.b64encode( b'0067462314' )
b'MDA2NzQ2MjMxNA=='
>>> 
~$ python2
Python 2.7.18 (default, Jul 14 2021, 08:11:37) 
[GCC 10.2.1 20210110] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import base64
>>> base64.b64encode( b'0067462314' )
'MDA2NzQ2MjMxNA=='
>>> 
~$ 
```
I'd say you have to ask THEM how the produce that result.

You could also try:   ;-)
```python
N = bytes((0x00, 0x67, 0x46, 0x23, 0x14))
>>> base64.b64encode(N)
b'AGdGIxQ='
```

**Reply by kjm1102** [reply]
Yeah looks like the b out the front isn't enough to indicate bytes we have to use the x notation
```
>>> base64.b64encode(b'\x00\x67\x45\x23\x14')
b'AGdFIxQ='
```

**Reply by kjm1102** [reply]
Meanwhile chat GPT is telling me otherwise
```
To convert the bytes representation b'0067462314' to the string representation "AGdGIxQ=" in Python, you can use the base64 module. Here's an example:

python
Copy code
import base64


bytes_value = b'0067462314'
string_value = base64.b64encode(bytes_value).decode('utf-8')

print(string_value)  # Output: AGdGIxQ=
In the code above, base64.b64encode(bytes_value) encodes the bytes value using Base64 encoding, and decode('utf-8') converts the resulting bytes to a string. The resulting string is then assigned to the variable string_value and printed to the console.
```

**Reply by robert-hh** [reply]
Obviously, ChatGPT is citing the wrong answer.

The difference between b'\x00\x67\x45\x23\x14' and b'0067462314'. The former represents 5 bytes, the latter 10 bytes.

**Reply by kjm1102** [reply]
Is there a way to get back to b'\x00\x67\x45\x23\x14' from 'AGdGIxQ=' Rob instead of  
```
>>> binascii.a2b_base64('AGdGIxQ=')
b'\x00gF#\x14'
```

**Reply by robert-hh** [reply]
The content of b'\x00gF#\x14' is the same as in b'\x00\x67\x45\x23\x14'. Python prints hex bytes as ASCII is possible. You can use b'\x00gF#\x14'.hex() to see the hex values, but without the \x inserts.

**Reply by kjm1102** [reply]
Are you doing a chat GPT impersonation Rob?
>>> b'\x00gF#\x14'.hex()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'bytes' object has no attribute 'hex'

**Reply by GitHubsSilverBullet** [reply]
Converting your original value forward and backward can be done in a single line.
Instead of chatting with a toy, I'd strongly suggest to read the documentation (both Python and MicroPython)
[https://docs.python.org/3.5/](url)
[https://docs.micropython.org/en/latest/](url)

```python
>>> from binascii import a2b_base64, b2a_base64, hexlify, unhexlify
>>> original = b'0067462314' 
>>> result1 = b2a_base64(unhexlify(original))
>>> result1
b'AGdGIxQ=\n'
>>> result2 = hexlify(a2b_base64(result1))
>>> result2
b'0067462314'
>>> result2 == original
True
```

**Reply by robert-hh** [reply]
> Are you doing a chat GPT impersonation Rob?

I usually try before posting. The hex() method for a bytes object exist e.g. on Python 3, and every board I tried. I do not know which platform you tried:
```
MicroPython v1.20.0-219-g47dc7d013-dirty on 2023-06-17; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> b'\x00gF#\x14'.hex()
'0067462314'
```

**Reply by kjm1102** [reply]
Ah tnx GHSB, that's what I couldn't figure out using my brain , how to swap backwards & forwards between the 2 representations, hence the desperate recourse to the AI after 2 hours of frustration.

Looks like I need to do a version upgrade if I want to use the elusive .hex() Rob
>>> import os; os.uname(); b'\x00gF#\x14'.hex()
(sysname='rp2', nodename='rp2', release='1.17.0', version='v1.0-pico-deepsleep-1-gaec7edbfc-dirty on 2021-11-06 (GNU 9.2.1 MinSizeRel)', machine='Raspberry Pi Pico (sleep) with RP2040')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'bytes' object has no attribute 'hex'

**Reply by jimmo** [reply]
`bytes.hex()` was added not long after v1.19 was released, and included in v1.20.

You are using v1.17 is nearly two years old now (!!!)

**Reply by kjm1102** [reply]
Yeah I know Jim. Thing is a guy fixed the pico deepsleep enough to be usable for me in v17 so I'm sort of stuck with it. It's a tough life for those of us not smart enough to do our own upython repairs.

**Comment by GitHubsSilverBullet** [comment]
> Meanwhile chat GPT is telling me otherwise
 
That's why it's called an AS system. (Artificial Stupidity)
Often it's wrong. Often it blatantly lies.
Takes more time to verify the questionable results it spews out, than to think about one's own solution. (Used to be called: learning)

Every once in a while it's OK to use one's own brain…
